README.md

```diff
@@ -138,8 +138,18 @@ appropriate remote recording support into your web app.
 ### Django
 `import appmap.django`. Adds `/_appmap/record` routes to a Django app.
 
-### Flask
-`import appmap.flask`. Adds `/_appmap/record` routes to a Flask app.
+## Flask
+```python
+from flask import Flask
```

Comment:
@apotterri 
```
from flask import Flask
from middleware import middleware

app = Flask('DemoApp')

# calling our middleware
app.wsgi_app = middleware(app.wsgi_app)
```
```
from flask import Flask

from appmap.flask import AppmapFlask

app = Flask(__name__)

appmap_flask = AppmapFlask(app)
```
The method of including middleware and extension is pretty much the same.

Flask middleware are wsgi middleware (intercept between wsgi and flask, notice wsgi_app passed to middleware instead of app), while extension could be called flask middleware. We would have to write handlers to intercept if the request is to `/_appmap/record', then handle GET, POST, DELETE and return custom response. As you can see extension code is pretty straightforward.

In Django, you have MIDDLEWARE in settings and BaseHandler.load_middleware loops through this setting and adds middleware.

Flask, AFAIK, doesn't have a middleware loader and you need to load it explicitly.

Reply:
Sorry I let this languish. It's been a pretty busy week....

Understood about the lack of built-in middleware loading. I wonder, though, if we can do something more "magical" if the project uses an application factory, and only require manual configuration if it doesn't. The less an application developer has to do to generate an AppMap, the better.

If we install ourselves as a [plugin](https://flask.palletsprojects.com/en/1.1.x/cli/#plugins), is there a way we can automatically add middleware? Maybe there's a builtin hook we can take advantage of, or we maybe we can wrap [`flask.cli.locate_app`](https://github.com/pallets/flask/blob/master/src/flask/cli.py#L251) and grab the app it returns.

Reply:
@apotterri based on your suggestion I was able to come up with this idea.

I am added the appmap as a command. If appmap is installed in the current environment, flask will automatically add a command `flask appmap`. This is just a wrapper around the default `flask run`. It enables APPMAP flag and starts a development server.

```
❯ flask                                                                                                                                                                                                                    ─╯
Usage: flask [OPTIONS] COMMAND [ARGS]...

  A general utility script for Flask applications.

  Provides commands from Flask, extensions, and the application. Loads the
  application defined in the FLASK_APP environment variable, or from a
  wsgi.py file. Setting the FLASK_ENV environment variable to 'development'
  will enable debug mode.

    $ export FLASK_APP=hello.py
    $ export FLASK_ENV=development
    $ flask run

Options:
  --version  Show the flask version
  --help     Show this message and exit.

Commands:
  appmap  Run a development server with appmap enabled
  routes  Show the routes for the app.
  run     Run a development server.
  shell   Run a shell in the app context.
```

```
❯ flask appmap --help                                                                                                                                                                                                      ─╯
Usage: flask appmap [OPTIONS]

Options:
  -h, --host TEXT                 The interface to bind to.
  -p, --port INTEGER              The port to bind to.
  --cert PATH                     Specify a certificate file to use HTTPS.
  --key FILE                      The key file to use when specifying a
                                  certificate.

  --reload / --no-reload          Enable or disable the reloader. By default
                                  the reloader is active if debug is enabled.

  --debugger / --no-debugger      Enable or disable the debugger. By default
                                  the debugger is active if debug is enabled.

  --eager-loading / --lazy-loader
                                  Enable or disable eager loading. By default
                                  eager loading is enabled if the reloader is
                                  disabled.

  --with-threads / --without-threads
                                  Enable or disable multithreading.
  --extra-files PATH              Extra files that trigger a reload on change.
                                  Multiple paths are separated by ':'.

  --help                          Show this message and exit.
```

```
❯ flask appmap                                                                                                                                                                                                             ─╯
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off
INFO:werkzeug: * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
INFO:werkzeug:127.0.0.1 - - [07/Mar/2021 12:33:54] "GET /_appmap/record HTTP/1.1" 200 -
```

```
❯ flask run                                                                                                                                                                                                                ─╯
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off
INFO:werkzeug: * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
INFO:werkzeug:127.0.0.1 - - [07/Mar/2021 12:34:26] "GET /_appmap/record HTTP/1.1" 404 -
```

Reply:
Thanks for this. I think we can make the integration tighter still. I opened https://github.com/virajkanwade/appmap-python/pull/1 with a somewhat different approach.